File: integration-tests/webdriver/src/main/java/com/gu/test/WebDriverFactory.java
package com.gu.test;

import static com.gu.test.PropertyLoader.getProperty;
import static org.apache.commons.lang3.StringUtils.isNotBlank;

import java.net.URL;
import java.util.concurrent.TimeUnit;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.remote.DesiredCapabilities;
import org.openqa.selenium.remote.RemoteWebDriver;

public class WebDriverFactory {

    private static final Log LOG = LogFactory.getLog(WebDriverFactory.class);
    public static final String BROWSER_VERSION = "30";
    public static final String OS_VERSION = "Windows 7";

    public static WebDriver createWebDriver() throws Exception {
        return createWebDriver(null);
    }

    public static WebDriver createWebDriver(String jobName) throws Exception {

        // Choose the browser, version, and platform to test
        DesiredCapabilities capabilities = DesiredCapabilities.firefox();
        capabilities.setCapability("version", BROWSER_VERSION);
        capabilities.setCapability("platform", OS_VERSION);
        if (isNotBlank(jobName)) {
            capabilities.setCapability("name", jobName);
        }

        String sauceLabsUrl = getProperty("saucelabs.remotedriver.url");
        // Create the connection to Sauce Labs to run the tests
        WebDriver driver = new RemoteWebDriver(new URL(sauceLabsUrl), capabilities);
        driver.manage().timeouts().implicitlyWait(30, TimeUnit.SECONDS);
        return driver;
    }
}

File: integration-tests/scala-fw-tests/src/main/scala/com/gu/integration/test/config/WebdriverFactory.scala
package com.gu.integration.test.config

import java.net.URL
import java.util.concurrent.TimeUnit.SECONDS
import org.apache.commons.lang3.StringUtils.isNotBlank
import org.openqa.selenium.Dimension
import org.openqa.selenium.Point
import org.openqa.selenium.WebDriver
import org.openqa.selenium.chrome.ChromeDriver
import org.openqa.selenium.firefox.FirefoxDriver
import org.openqa.selenium.ie.InternetExplorerDriver
import org.openqa.selenium.remote.DesiredCapabilities
import org.openqa.selenium.remote.RemoteWebDriver
import com.gu.integration.test.config.PropertyLoader._
import com.gu.integration.test.config.PropertyLoader.getProperty
import com.gu.automation.support.TestLogging

object WebdriverFactory extends TestLogging {
  val Firefox = "firefox"
  val Chrome = "chrome"
  val Ie = "ie"

  def createWebDriver(testCaseName: String): WebDriver = {
    val sauceLabsUrl = getProperty(SauceLabsUrl)
    val browserName = getProperty(Browser)
    
    val capabilities = browserName match {
      case Firefox => initDesiredCapabilities(testCaseName, DesiredCapabilities.firefox())
      case Chrome => initDesiredCapabilities(testCaseName, DesiredCapabilities.chrome())
      case Ie => initDesiredCapabilities(testCaseName, DesiredCapabilities.internetExplorer())
      case unknown => throw new RuntimeException(s"Browser: [$unknown] is not supported")
    }
    
    val webDriver = if (isNotBlank(sauceLabsUrl)) {
      new RemoteWebDriver(new URL(sauceLabsUrl), capabilities)
    } else browserName match {
      case Firefox => new FirefoxDriver(capabilities)
      case Chrome => new ChromeDriver(capabilities)
      case Ie => new InternetExplorerDriver(capabilities)
    }

    initWebDriver(webDriver)
  }

  private def initWebDriver(webDriver: WebDriver): WebDriver = {
    webDriver.manage().window().setPosition(new Point(0, 0))
    //needs to set a custom size because the auto size sometimes is not large enough to display certain elements
    webDriver.manage().window().setSize(new Dimension(1600, 1024))
    webDriver.manage().timeouts().implicitlyWait(10, SECONDS)
    webDriver
  }

  private def initDesiredCapabilities(testCaseName: String, capabilities: DesiredCapabilities): DesiredCapabilities = {
    capabilities.setCapability("version", getProperty(BrowserVersion))
    if (isNotBlank(testCaseName)) {
      capabilities.setCapability("name", testCaseName)
    }
    return capabilities
  }
}
